packages/sqlite/src/index.ts

```diff
@@ -118,115 +92,91 @@ class DatabaseConnectionImplementation implements DatabaseConnection {
 
   async dispose() {
-    await new Promise<void>((resolve, reject) => {
-      this._database.close((err) => {
-        if (err) reject(err);
-        else resolve();
-      });
-    });
+    this._database.close();
   }
 }
 
 export default function connect(
   filename: string = IN_MEMORY,
-  options: DatabaseConnectionOptions = {},
+  options: DatabaseOptions = {},
 ): DatabaseConnection {
   return new DatabaseConnectionImplementation(filename, options);
 }
 module.exports = Object.assign(connect, {
   default: connect,
-  DatabaseConnectionMode,
   IN_MEMORY,
   sql,
   isSqlQuery,
 });
 
 async function runQuery(
   query: SQLQuery,
-  database: sqlite.Database,
+  database: Database,
   lock: <T>(fn: () => Promise<T>) => Promise<T>,
 ) {
   const {text, values} = query.format(sqliteFormat);
   return lock(async () => {
-    return await new Promise<any[]>((resolve, reject) => {
-      database.all(text, values, (err, rows) => {
-        if (err) reject(err);
-        else resolve(rows);
-      });
-    });
+    const stm = database.prepare(text);
+    try {
+      return stm.all(...values);
+    } catch (_err) {
+      const err = _err as Error;
+      // TODO we should be able to catch this before calling all()
+      if (err.message.indexOf('This statement does not return data') >= 0) {
+        return stm.run(...values);
+      }
+      throw err
+    }
   });
 }
 
-interface Queue<T> {
-  push(item: T): void;
-  pop(): Promise<T>;
-  /**
-   * Amount of items in the queue
-   * This can be negative if pop has been called more times than push.
-   */
-  length: number;
-}
 async function* runQueryStream(
   query: SQLQuery,
-  database: sqlite.Database,
+  database: Database,
   lock: <T>(fn: () => Promise<T>) => Promise<T>,
 ): AsyncIterableIterator<any> {
-  const queue: Queue<{done: false; value: any} | {done: true; err: any}> =
-    new Queue();
   const {text, values} = query.format(sqliteFormat);
+  // TODO figure out how to avoid the extra closures
+  let resolve = () => {};
+  let reject: (reason? : any) => void;
+  reject = () => {};
+  const promise = new Promise<void>((res, rej) => {
+    resolve = res
+    reject = rej
+  })
+  let rows
   lock(async () => {
-    await new Promise<void>((releaseMutex) => {
-      database.each(
-        text,
-        values,
-        (err, row) => {
-          if (err) queue.push({done: true, err});
-          else queue.push({done: false, value: row});
-        },
-        (err) => {
-          releaseMutex();
-          queue.push({done: true, err});
-        },
-      );
-    });
+    const stm = database.prepare(text);
+    rows = stm.iterate(...values);
+    await promise
   }).catch((ex) => {
     setImmediate(() => {
       throw ex;
     });
   });
-  let value = await queue.pop();
-  while (!value.done) {
-    yield value.value;
-    value = await queue.pop();
+  try {
+    for (const row of (rows || [])) {
+      yield row;
+    }
+    resolve()
+  } finally {
+    reject()
```

Comment:
You would need to reject with an error explaining what went wrong. I think in this instance it may be ok to just always `resolve the promise to release the lock:

```suggestion
  } finally {
    resolve()
```